src/main/kotlin/com/nikitanazarov/coroutinestacks/CoroutineTraceForestBuilder.kt

```diff
@@ -0,0 +1,173 @@
+        vertex.addListSelectionListener { e ->
+            val currentList = e.source as? JBList<*> ?: return@addListSelectionListener
+            if (previousListSelection != currentList) {
+                previousListSelection?.clearSelection()
+            }
+            previousListSelection = currentList
+        }
+        vertexData.add(vertex)
+        maxWidth += vertex.preferredWidth
+        traceNotNullCount += 1
+    }
+
+    if (traceNotNullCount == 0) {
+        return null
+    }
+    val averagePreferredWidth = maxWidth / traceNotNullCount
+    vertexData.forEach { vertex ->
+        if (vertex != null) {
+            vertex.preferredWidth = averagePreferredWidth
+            componentData.add(vertex)
+            return@forEach
+        }
+        componentData.add(Separator())
+    }
+
+    val forest = ContainerWithEdges()
+    componentData.forEach { forest.add(it) }
+    forest.layout = ForestLayout()
+    val panel = DraggablePanel()
+    panel.add(forest)
+
+    return JBScrollPane(panel)
+}
+
+private fun createCoroutineTraces(rootValue: Node): List<CoroutineTrace?> {
+    val stack = Stack<Pair<Node, Int>>().apply { push(rootValue to 0) }
+    val parentStack = Stack<Node>()
+    var previousLevel: Int? = null
+    val coroutineTraces = mutableListOf<CoroutineTrace?>()
+
+    while (stack.isNotEmpty()) {
+        val (currentNode, currentLevel) = stack.pop()
+        val parent = if (parentStack.isNotEmpty()) parentStack.pop() else null
+        var coroutineStackHeader = CoroutineStacksBundle.message("number.of.coroutine")
+
+        if (parent != null && parent.num != currentNode.num) {
+            if (currentNode.num > 1)
+                coroutineStackHeader = CoroutineStacksBundle.message("number.of.coroutines", currentNode.num)
+            val currentTrace = CoroutineTrace(
+                mutableListOf(currentNode.stackFrameItem),
+                coroutineStackHeader,
+                currentNode.coroutinesActive
+            )
+            repeat((previousLevel ?: 0) - currentLevel + 1) {
+                coroutineTraces.add(null)
+            }
+            coroutineTraces.add(currentTrace)
+            previousLevel = currentLevel
+        } else if (parent != null) {
+            coroutineTraces.lastOrNull()?.stackFrameItems?.add(0, currentNode.stackFrameItem)
+        }
+
+        currentNode.children.values.reversed().forEach { child ->
+            stack.push(child to (currentLevel + 1))
+            parentStack.push(currentNode)
+        }
+    }
+
+    return coroutineTraces
+}
+
+private fun SuspendContextImpl.buildStackFrameGraph(
+    coroutineDataList: List<CoroutineInfoData>,
+    rootValue: Node
+): String {
+    var lastRunningStackFrame = ""
+
+    coroutineDataList.forEach { coroutineData ->
+        var currentNode = rootValue
+
+        val coroutineFrameItemLists = CoroutineFrameBuilder.build(coroutineData, this)
```

Comment:
`val coroutineFrameItemLists = CoroutineFrameBuilder.build(coroutineData, this) ?: return@forEach`